src/test/app/AMM_test.cpp

```diff
@@ -2986,6 +2985,208 @@ struct AMM_test : public jtx::AMMTest
+
+                BEAST_EXPECT(std::apply(
+                    [&](STAmount pool1, STAmount pool2, STAmount lpToken) {
+                        return pool1 * pool2 >= (lpToken * lpToken);
+                    },
+                    ammAlice.balances(XRP, USD)));
+
+                // payment transaction: sendmax is the parameterised value
+                env(pay(bob, carol, STAmount{USD, 1, 7}),
+                    sendmax(money),
+                    txflags(tfPartialPayment));
+                env.close();
+
+                BEAST_EXPECT(std::apply(
+                    [&](STAmount pool1, STAmount pool2, STAmount lpToken) {
+                        return pool1 * pool2 >= (lpToken * lpToken);
+                    },
+                    ammAlice.balances(XRP, USD)));
+
+                // withdraw transaction
+                ammAlice.withdraw(
+                    WithdrawArg{.account = bob, .asset1Out = money});
+
+                BEAST_EXPECT(std::apply(
+                    [&](STAmount pool1, STAmount pool2, STAmount lpToken) {
+                        return pool1 * pool2 >= (lpToken * lpToken);
+                    },
+                    ammAlice.balances(XRP, USD)));
+            });
+        }
+
+        // the below transactions use the IOU currency to perform deposit,
+        // payment and withdrawl transactions, against an XRP/IOU AMM pool
+        for (STAmount const& money :
+             {STAmount(USD, 1, -3), STAmount(USD, 1, 0), STAmount(USD, 1, 6)})
+        {
+            testAMM([&](AMM& ammAlice, Env& env) {
+                fund(env, gw, {bob}, XRP(1'000), {USD(10'000'000)}, Fund::Acct);
+                env.close();
+
+                // Note: The correct-ness of the balances post-payment,
+                // deposit and withdraw transactions has not been validated.
+                // This has been already validated elsewhere in other unit tests
+                // Rather, the AMM Invariant condition is checked after every
+                // relevant transaction.
+
+                // Note: Despite depositing USD (which is the second asset in
+                // the AMM pool), unit tests use a convention of asset1In to
+                // denote single-asset deposits. It is a similar case with the
+                // AMMWithdraw transaction.
+                ammAlice.deposit(DepositArg{.account = bob, .asset1In = money});
+
+                BEAST_EXPECT(std::apply(
+                    [&](STAmount pool1, STAmount pool2, STAmount lpToken) {
+                        return pool1 * pool2 >= (lpToken * lpToken);
+                    },
+                    ammAlice.balances(XRP, USD)));
+
+                // payment transaction: sendmax is the parameterised value
+                env(pay(bob, carol, STAmount{XRP, 1, 7}),
+                    sendmax(money),
+                    txflags(tfPartialPayment));
+                env.close();
+
+                BEAST_EXPECT(std::apply(
```

Comment:
If anyone has any ideas on how I can remove this ugly repetitive invariant checks, please let me know.

I'd like to perform the AMMInvariant check after every transaction. I couldn't get `BEAST_EXPECT(std::apply(AMMInvariantCheck, ammAlice.balances(XRP, USD)));` this to work.
The compiler complains about the need to use a non-static callable inside `std::apply` ://